### PR TITLE
Trivial support for ipv6

### PR DESCRIPTION
**Pull request by raul-arabaolaza**
Needed by some proprietary code I believe this could also be interesting to the community.
While arguably we want to stop using this library and move to test containers this change is so small that I can not resist to contribute it back. There may be a potential effort on migrating the ath and plugin tests to test containers that I totally endorse, but nothing prevents us to do that and this in parallel given the test containers migration is gonna require more effort.
  

### Testing done

```[tasklist]
### Submitter checklist
- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

**Comment by raul-arabaolaza** [comment]
Successfully tested on proprietary runs of both the ATH and the PCT

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java [hunk 1]
```diff
@@ -90,9 +90,10 @@ public String ipBound(int n) {
                 return getIpAddress();
             }
             String out = Docker.cmd("port").add(cid, n).popen().verifyOrDieWith("docker port command failed").trim();
-            if (out.isEmpty())  // expected to return single line like "0.0.0.0:55326"
+            if (out.isEmpty())  // expected to return single line like "0.0.0.0:55326" or [::]:55326
                 throw new IllegalStateException(format("Port %d is not mapped for container %s", n, cid));
-            return out.split(":")[0];
+            // assumes it is published only in ipv6 or only in ipv4
+            return ipv6Enabled() ? out.substring(1,out.lastIndexOf(":") - 1) : out.split(":")[0];
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java:96]: Why would this need to be conditional? AFAICT both IPv4 and IPv6 formats would work fine by taking the last `:` to be the port suffix, and anything before that to be an IP address.

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java [hunk 2]
```diff
@@ -108,9 +109,9 @@ public String ipUdpBound(int n) {
                 return getIpAddress();
             }
             String out = Docker.cmd("port").add(cid, n + "/udp").popen().verifyOrDieWith("docker port command failed").trim();
-            if (out.isEmpty())  // expected to return single line like "0.0.0.0:55326"
+            if (out.isEmpty())  // expected to return single line like "0.0.0.0:55326" or [::]:55326
                 throw new IllegalStateException(format("Udp port %d is not mapped for container %s", n, cid));
-            return out.split(":")[0];
+            return ipv6Enabled() ? out.substring(1,out.lastIndexOf(":") -1) : out.split(":")[0];
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java:114]: ditto

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java [hunk 3]
```diff
@@ -126,10 +127,10 @@ public int port(int n) {
                 return n;
             }
             String out = Docker.cmd("port").add(cid, n).popen().verifyOrDieWith("docker port command failed").trim();
-            if (out.isEmpty())  // expected to return single line like "0.0.0.0:55326"
+            if (out.isEmpty())  // expected to return single line like "0.0.0.0:55326" or [::]:55326
                 throw new IllegalStateException(format("Port %d is not mapped for container %s", n, cid));
 
-            return Integer.parseInt(out.split(":")[1]);
+            return Integer.parseInt(out.split(":")[ipv6Enabled() ? 3 : 1]);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java:133]: ditto

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java [hunk 4]
```diff
@@ -145,10 +146,10 @@ public int udpPort(int n) {
                 return n;
             }
             String out = Docker.cmd("port").add(cid, n + "/udp").popen().verifyOrDieWith("docker port command failed").trim();
-            if (out.isEmpty())  // expected to return single line like "0.0.0.0:55326"
+            if (out.isEmpty())  // expected to return single line like "0.0.0.0:55326" or [::]:55326
                 throw new IllegalStateException(format("Udp port %d is not mapped for container %s", n, cid));
 
-            return Integer.parseInt(out.split(":")[1]);
+            return Integer.parseInt(out.split(":")[ipv6Enabled() ? 3 : 1]);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java:152]: ditto

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java [hunk 5]
```diff
@@ -233,4 +235,12 @@ public String toString() {
     public boolean sharingHostDockerService() {
         return Boolean.valueOf(System.getenv("SHARED_DOCKER_SERVICE"));
     }
+
+    public static boolean ipv6Enabled() {
+        return Boolean.getBoolean("java.net.preferIPv6Addresses");
+    }
+
+    public static String encloseInBrackets(String toEnclose) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java:243]: trivial, inline

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java (hunk 1 above)
Comment (raul-arabaolaza) [reply src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java]: Also need to remove brackets if present as `docker port` output rightfully encloses ipv6 ips in brackets but you are right the conditional can be removed, changed in be883c6824d80b5bcbf8e4244369c9d5bbb708cb

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java (hunk 2 above)
Comment (raul-arabaolaza) [reply src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java]: be883c6824d80b5bcbf8e4244369c9d5bbb708cb

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java (hunk 3 above)
Comment (raul-arabaolaza) [reply src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java]: be883c6824d80b5bcbf8e4244369c9d5bbb708cb

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java (hunk 4 above)
Comment (raul-arabaolaza) [reply src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java]: be883c6824d80b5bcbf8e4244369c9d5bbb708cb

File: src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java (hunk 5 above)
Comment (raul-arabaolaza) [reply src/main/java/org/jenkinsci/test/acceptance/docker/DockerContainer.java]: be883c6824d80b5bcbf8e4244369c9d5bbb708cb

**Review by jglick — APPROVED** [review]

**Pull request merged**